Add `jl_print_task_backtraces()`

kpamnany opened the pull request:
Iterates through `jl_all_tls_states` and through all `live_tasks` in `ptls->heap`, printing backtraces.

The purpose is to help find deadlocks.

Replaces https://github.com/JuliaLang/julia/pull/44990.
Closes #46177.

kpamnany committed:
Add `jl_print_task_backtraces()`

Iterates through `jl_all_tls_states` and through all `live_tasks`
in `ptls->heap`, printing backtraces.

vilterp commented on src/stackwalk.c:
```diff
@@ -1108,6 +1108,42 @@ JL_DLLEXPORT void jl_print_backtrace(void) JL_NOTSAFEPOINT
 }
 
+JL_DLLEXPORT void jl_print_task_backtraces(void)
+{
```
It might be helpful for this to take an IO stream or an in-memory buffer as an argument, so we could direct its output to e.g. a a file or HTTP response.

Not sure if that would create locking/scheduling complications.

The Go version takes a buffer as an argument, maybe for this reason: https://pkg.go.dev/runtime#Stack

NHDaly replied:
That's a good idea, yeah!

kpamnany replied:
This would need more surgery (the actual printing of backtrace entries is [a](https://github.com/JuliaLang/julia/blob/master/src/stackwalk.c#L1102) [few](https://github.com/JuliaLang/julia/blob/master/src/stackwalk.c#L657) [levels](https://github.com/JuliaLang/julia/blob/master/src/stackwalk.c#L645) [down](https://github.com/JuliaLang/julia/blob/master/src/stackwalk.c#L620)) and maybe changing the interface for the existing functions (`jl_print_backtrace()`, `jlbacktracet()`, ...).

IMO, that should be a separate PR.

NHDaly commented on src/stackwalk.c:
```diff
@@ -1108,6 +1108,42 @@ JL_DLLEXPORT void jl_print_backtrace(void) JL_NOTSAFEPOINT
 
+JL_DLLEXPORT void jl_print_task_backtraces(void)
+{
+    jl_task_t *t;
+    jl_ptls_t ptls;
+    arraylist_t *live_tasks;
+    size_t i, j, n;
```
nit:
If i'm not mistaken, in the later C standards (C11 onward), which I think we use, you no longer need to declare the variables' types at the top of the function.

I think the preference is (inherited from C++) to keep variable declaration as tightly-scoped as possible, which usually means putting the type declaration on the first assignment.

Does that sound right to you?

vilterp reviewed (commented):
Nice!

NHDaly commented on src/stackwalk.c:
```diff
@@ -1108,6 +1108,42 @@ JL_DLLEXPORT void jl_print_backtrace(void) JL_NOTSAFEPOINT
+        ptls = jl_all_tls_states[i];
+        live_tasks = &ptls->heap.live_tasks;
+        n = live_tasks->len;
+        jl_safe_printf("==== Thread %d (%zu live tasks)\n",
+                ptls->tid + 1, n + 1);
```
Heh, i've already forgotten again since JuliaCon why exactly this array is per-thread, since tasks can of course move threads, and when the task is blocking, it's not scheduled on any thread...

Does this correlate to which Thread _created_ the task? Could we maybe clarify the relationship in the println? (For example `==== Tasks created by Thread %d` or something like that?)

kpamnany replied:
A task is put onto `ptls->heap.live_tasks` [here](https://github.com/JuliaLang/julia/blob/master/src/gc-stacks.c#L179), when a stack is allocated for it. This usually happens when in the creating task's context and hence on the thread that was running the creating task but in some situations, this can happen when the created task first runs, which could be on any thread.

From a point of view, this partitioning of tasks across threads is not meaningful. But I can add that clarification in the output anyway.

NHDaly replied:
Thanks, yeah i think that will be helpful!

NHDaly commented on src/stackwalk.c:
```diff
@@ -1108,6 +1108,42 @@ JL_DLLEXPORT void jl_print_backtrace(void) JL_NOTSAFEPOINT
+                jl_safe_printf("      no stack\n");
+        }
+    }
+    jl_safe_printf("==== Done\n");
```
```suggestion
    jl_safe_printf("==== End of Thread %d\n", ptls->tid + 1);
```
🤷 i think this might make the log easier to read if there's a longgggg gap between the start and end?

NHDaly commented on src/stackwalk.c:
```diff
@@ -1108,6 +1108,42 @@ JL_DLLEXPORT void jl_print_backtrace(void) JL_NOTSAFEPOINT
+        jl_safe_printf("          (sticky: %d, started: %d, state: %d, tid: %d)\n",
+                t->sticky, t->started, jl_atomic_load_relaxed(&t->_state),
+                jl_atomic_load_relaxed(&t->tid) + 1);
+        jlbacktracet(t);
```
Will it ever be possible for the root task's sttkbuf to be NULL? I think probably not, but is it worth guarding against it, here?

kpamnany replied:
`jlbacktracet(t)` calls `jl_rec_backtrace()` which [checks](https://github.com/JuliaLang/julia/blob/master/src/stackwalk.c#L866-L867).

NHDaly reviewed (approved):
Excellent, this looks super helpful. Thank you @kpamnany!!

Can we also backport to 1.8? 🥺 I think this would be quite helpful for debugging issues on 1.8 too, if we can.

NHDaly commented on src/stackwalk.c:
```diff
@@ -1108,6 +1108,42 @@ JL_DLLEXPORT void jl_print_backtrace(void) JL_NOTSAFEPOINT
+    }
+    jl_safe_printf("==== Done\n");
+}
+
```
Oh, actually, can you please add tests? At least one that exercises the functionality.

kpamnany replied:
There are no tests for any of the printing functions in `stackwalk.c`. I could add one/some, but am dubious of the value added.

NHDaly replied:
it seems that it's useful to have at least one test that just calls the function to exercise it, and make sure that it doesn't crash..
But i agree that since this is C, not julia, at least you know it compiles, so a test is slightly less necessary.

vtjnash commented on src/stackwalk.c:
```diff
@@ -1108,6 +1108,42 @@ JL_DLLEXPORT void jl_print_backtrace(void) JL_NOTSAFEPOINT
+    size_t i, j, n;
+
+    for (i = 0; i < jl_n_threads; i++) {
+        ptls = jl_all_tls_states[i];
```
traditionally, we call this `ptls2` in gC, but you also aren't allowed to access that outside of GC, since it is thread-local state for a different thread.

kpamnany replied:
Yes, it will be racey, but will it simply fail on some platforms?

kpamnany replied:
If, on some platforms, it is not possible to access another thread's space through `jl_all_tls_states`, the only way around I can see is to do a GC stop-the-world and have each thread print its own `live_tasks` which will no longer be racy, but make this a lot more complicated and slow.

vtjnash replied:
A thread, in general, cannot print the stack of another thread's task either, so having it run behind the `jl_gc_wait_for_the_world` logic (running on every thread) would be useful. Though deadlock reporting doesn't precisely need that, since that may be only essential to run if we have run out of all other useful work and halted the other threads anyways.

vtjnash replied:
The `live_tasks` pointer may move from under you while you are looking at it, making this very dangerous. You are right that sometimes we examine values in the `jl_all_tls_states`, but only if they are marked atomic or are protected by a lock

NHDaly replied:
As far as I know, we only intended to call this from `gdb`, in which case the world will already be artificially stopped.

I agree with you both, that I don't think this is safe to call from inside the julia runtime.

I think we should make sure that the function comments make it clear that this isn't safe to call while the runtime is running.
It might make sense to provide another version of it, which force-pauses all the threads before printing, reusing the GC machinery, but that seems like a follow up PR?

vilterp replied:
FWIW the Go version of this stops the world so it can be safely called from 'userspace': https://github.com/golang/go/blob/6d8ec893039a39f495c8139012e47754e4518b70/src/runtime/mprof.go#L1193-L1196

This is sometimes put in an HTTP handler for easy access; would be nice for us to eventually be able to do the same. GDB is a good start ofc :)

kpamnany commented:
For this error:
```
/cache/build/default-amdci5-6/julialang/julia-master/src/stackwalk.c:1124:9: error: Passing non-rooted value as argument to function that may GC [julia.GCChecker]
--
  | jlbacktracet(ptls2->root_task);
  | ^            ~~~~~~~~~~~~~~~~
```
Do I have to `JL_GC_PUSH1(&ptls2->root_task)` and `JL_GC_POP()` after?

kpamnany commented:
Also, I don't understand this?
```
/cache/build/default-amdci5-6/julialang/julia-master/src/stackwalk.c:1119:54: note: Argument value was derived global with untracked type. You may want to update the checker's type list
--
  | jl_safe_printf("     ---- Root task (%p)\n", ptls2->root_task);
  | ^~~~~~~~~~~~~~~~
```

kpamnany commented:
We're experimenting with this to see if it's useful. As Nathan said, this is basically for use in `gdb`, i.e. all threads will be stopped, so poking into other threads' local storage is safe.

I find that I cannot get a backtrace (Linux glibc x86-64) with the default `JL_HAVE_ASM` and have to turn on `JL_HAVE_UNW_CONTEXT`. Does https://github.com/JuliaLang/julia/pull/45110 fix that @vtjnash?

vtjnash commented:
This should be useful eventually for figuring out why the Sockets test stales on CI, such as https://buildkite.com/julialang/julia-master/builds/16200#01838101-13bb-48fd-8aff-7705c619cd66

NHDaly commented:
> I find that I cannot get a backtrace (Linux glibc x86-64) with the default `JL_HAVE_ASM` and have to turn on `JL_HAVE_UNW_CONTEXT`. Does #45110 fix that @vtjnash?

Oh, interesting! I wonder if that's why I originally was thinking this didn't work. Thanks for tracking all of this down, @kpamnany!

@vtjnash / @kpamnany: is this resolved? Is there anything I/we can do to help move this along? Thanks! 😊

kpamnany commented:
I added a comment warning that this is only intended for use when all threads are stopped (i.e. in `gdb`). I also removed it from exported functions for that reason.

We've verified that this can be useful. So, apart from the GC checker errors, I think this is good to go.

NHDaly reviewed (approved):
👍 LGTM thanks @kpamnany!

vchuravy reviewed (approved):
LGTM! I would like a finally comment by @vtjnash saying if he is okay with the current intended usage (only from GDB) and then a follow-up PR that would implement Go solution to briefly stop-the-world.

Keno commented:
This is failing analyzegc on master. Not sure why it didn't fail on this PR:
```
/cache/build/default-amdci5-5/julialang/julia-master/src/stackwalk.c:1131:28: error: Implicit Atomic seq_cst synchronization [concurrency-implicit-atomics,-warnings-as-errors]
--
  | for (size_t i = 0; i < jl_n_threads; i++) {
  | ^
  | /cache/build/default-amdci5-5/julialang/julia-master/src/stackwalk.c:1132:27: error: Implicit Atomic seq_cst synchronization [concurrency-implicit-atomics,-warnings-as-errors]
  | jl_ptls_t ptls2 = jl_all_tls_states[i];
  | ^
```

giordano commented:
This PR was reverted in #47182 because of the problem with analyzegc reported above.
